lib/client/keyagent_test.go

```diff
@@ -452,39 +454,60 @@ func (s *KeyAgentTestSuite) makeKey(username string, allowedLogins []string, ttl
 	}, nil
 }
 
-func startDebugAgent() error {
-	errorC := make(chan error)
+func startDebugAgent() (closer func(), err error) {
+	startedC := make(chan struct{})
 	rand.Seed(time.Now().Unix())
 
-	go func() {
-		socketpath := filepath.Join(os.TempDir(),
-			fmt.Sprintf("teleport-%d-%d.socket", os.Getpid(), rand.Uint32()))
-
-		systemAgent := agent.NewKeyring()
-
-		listener, err := net.Listen("unix", socketpath)
-		if err != nil {
-			errorC <- trace.Wrap(err)
-			return
-		}
-		defer listener.Close()
+	socketpath := filepath.Join(os.TempDir(),
+		fmt.Sprintf("teleport-%d-%d.socket", os.Getpid(), rand.Uint32()))
 
-		os.Setenv(teleport.SSHAuthSock, socketpath)
+	listener, err := net.Listen("unix", socketpath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-		// agent is listeninging and environment variable is set unblock now
-		close(errorC)
+	systemAgent := agent.NewKeyring()
+	os.Setenv(teleport.SSHAuthSock, socketpath)
 
+	doneC := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		defer os.Setenv(teleport.SSHAuthSock, "")
+		// agent is listening and environment variable is set, unblock now
+		close(startedC)
 		for {
 			conn, err := listener.Accept()
 			if err != nil {
-				log.Warnf("Unexpected response from listener.Accept: %v", err)
-				continue
+				if !utils.IsUseOfClosedNetworkError(err) {
+					log.Warnf("Unexpected response from listener.Accept: %v", err)
+				}
+				return
 			}
-
-			go agent.ServeAgent(systemAgent, conn)
+			wg.Add(2)
+			go func() {
+				agent.ServeAgent(systemAgent, conn)
+				wg.Done()
+			}()
+			go func() {
+				<-doneC
+				conn.Close()
+				wg.Done()
+			}()
 		}
 	}()
 
+	go func() {
+		<-doneC
+		listener.Close()
+		wg.Done()
+	}()
+
 	// block until agent is started
-	return <-errorC
+	<-startedC
```

Comment:
Should this use select with a context to allow for a timeout?

Reply:
The test run as a whole has a timeout - if this is a hard block, it will be forcibly terminated (10min by default). Otherwise, I feel that using any other timeout is arbitrary and only gets you this far. E.g. during development you can limit the time on command line, the automatic tests will also report failure if anything.